README.md
# iButton to Wiegand

Načítanie kódu z kľúča iButton a jeho prevod na požadovaný protokol Wiegand 26, 34 alebo 42.



CHANGELOG.md
# Changelog

Všetky zmeny, ktoré stoja za zmienku budú zverejnené v tomto súbore.

Formát je založený na [Udržuj Changelog] (https://keepachangelog.com/sk/1.0.0/),
a tento projekt dodržiava [Sémantické verzovanie] (https://semver.org/lang/sk/spec/v2.0.0.html).

Typy zmien
    Added - pre nové funkcie.
    Changed - pre zmeny existujúcej funkcie.
    Deprecated - pre funkcie, ktoré budú čoskoro odstránené.
    Removed - pre odstránené funkcie.
    Fixed - pre opravy chýb.
    Security - v prípade bezpečnostných zraniteľností.


## [Unreleased]

## [3.0.2] - 2019-09-20
### Added
- Doplnenie súboru CHANGELOG.md
### Changed
- Úpravy výpisov hexadecimálnych čísiel
- Úprava popisu programu v súbore README.md

## [3.0.1] - 2017-05-12
### Changed
- Zmena zapojenia pinov
- Skrátenie doby spánku

## [3.0.0] - 2017-03-27
### Added
- Doplnenie protokolu Wiegand 34
### Changed
- Spracovanie protokolu v oddelených súboroch
- Úprava identifikácie na základe tzv. Family Code

## [2.1.0] - 2017-03-24

## [2.0.0] - 2017-03

## [1.0.0] - 2016



Wiegand.h
#ifndef WIEGAND_H
#define WIEGAND_H

#include <Arduino.h>
#include "FamilyCode.h"

#define DEBUG 1

// Definicie protokolu
const uint8_t WIEGAND26 = 3;  // pocet bajtov predstavujucich data
const uint8_t WIEGAND34 = 4;
const uint8_t WIEGAND42 = 5;

// Struktura pre urcenie vtahu vstupov a vystupov
struct Slot {
  uint8_t citacka;    // vstupny pin pre pripojenie citacky
  uint8_t vystup[2];  // vystupne piny - D0 (index 0), D1 (index 1)
  uint8_t protokol;
};

// Struktura paketu protokolu Wiegand 26/34/42
struct Paket {
  uint8_t paritaMSB;
  uint8_t *data;
  uint8_t paritaLSB;
};


// Deklaracia funkcii
extern void posliKod( const uint8_t * k, uint8_t p, const uint8_t * v );

static void vytvorPaket( uint8_t p, struct Paket & pkt );
//static uint8_t vypocitajParitu( uint32_t n );
static uint8_t vypocitajParitu( uint8_t n );
static void posliBit( uint8_t vystup);

#endif  // #ifndef WIEGAND_H


Wiegand.cpp
#include "Wiegand.h"

uint8_t i;

/*******************************************************************************
 *    Function: posliKod
 * Description: Odoslanie kodu v prislusnom protokole
 *   Parameter: [IN] const uint8_t * k - unikatny ROM kod
 *              [IN] uint8_t p - typ protokolu - Wiegand 26/34/42
 *              [IN] uint8_t v - pole vystupov
 *      Return: nič
 ******************************************************************************/
void posliKod( const uint8_t *k, uint8_t p, const uint8_t * v )
{
  struct Paket paket = { 0, k + 1, 0 };
  vytvorPaket( p, paket );

  #if DEBUG == 1
    Serial.print( F( "Paritny bit MSB: " ));
    Serial.println( paket.paritaMSB );
    Serial.println( F( "KOD: " ));
    for ( i = 0; i < p; i++ ) {
      if ( paket.data[ i ] < 0x10 )
        Serial.print( "0" );
      Serial.print( paket.data[ i ], HEX );
    }
    Serial.println();
    //uint32_t dec = 0;
    //for ( i = 0; i < p; i++ )
    //  dec |= paket.data[ i ] << ( 8 * ( p - 1 - i ));
    //Serial.println( dec );
    for ( i = 0; i < p; i++ )
      Serial.print( paket.data[ i ], BIN );
    Serial.println();
    Serial.print( F( "Paritny bit LSB: " ));
    Serial.println( paket.paritaLSB );
  #endif
  posliBit( v[ paket.paritaMSB ]);
    for ( i = 0; i < p; i++) {
      for ( int8_t b = 7; b >= 0; b-- ) {
        posliBit( v[ ( paket.data[ i ] >> b ) & 0x01 ] );
      }
    }
  posliBit( v[ paket.paritaLSB ]);
}

/*******************************************************************************
 *    Function: vytvorPaket
 * Description: Vytvorenie paketu protokolu Wiegand
 *   Parameter: [IN] const uint8_t p - typ protokolu - Wiegand 26/34/42
 *              [OUT] struct Paket * pkt - vytvorenie paketu protokolu
 *      Return: nič
 ******************************************************************************/
static void vytvorPaket( uint8_t p, struct Paket & pkt )
{
  switch ( p ) {
    case WIEGAND26: {
      // Spocitanie parity pre hornych 12 bitov
      // Parny paritny bit je nastaveny na 1, ak je pocet jednotiek v datovom slove neparny
      pkt.paritaMSB = ( vypocitajParitu( pkt.data[ 0 ]) + vypocitajParitu( pkt.data[ 1 ] & 0xF0 )) & 0x01;
      
      // Spocitanie parity pre dolnych 12 bitov
      // Neparny paritny bit je nastaveny na 1, ak je pocet jednotiek v datovom slove parny
      pkt.paritaLSB = ~( vypocitajParitu( pkt.data[ 1 ] & 0x0F) + vypocitajParitu( pkt.data[ 2 ]) ) & 0x01;
    } break;
    case WIEGAND34: {
      // Spocitanie parity pre hornych 16 bitov
      // Parny paritny bit je nastaveny na 1, ak je pocet jednotiek v datovom slove neparny
      pkt.paritaMSB = ( vypocitajParitu( pkt.data[ 0 ]) + vypocitajParitu( pkt.data[ 1 ]) ) & 0x01;
      
      // Spocitanie parity pre dolnych 16 bitov
      // Neparny paritny bit je nastaveny na 1, ak je pocet jednotiek v datovom slove parny
      pkt.paritaLSB = ~( vypocitajParitu( pkt.data[ 2 ]) + vypocitajParitu( pkt.data[ 3 ]) ) & 0x01;
    } break;
    case WIEGAND42: {
      // Paritny bit nastaveny na 0
      pkt.paritaMSB = 0;
      
      // Spocitanie parity pre vsetkych 40 bitov
      // Parny paritny bit je nastaveny na 1, ak je pocet jednotiek v datovom slove neparny
      for ( i = 0; i < p; i++ )
        pkt.paritaLSB += vypocitajParitu( pkt.data[ i ]);
      pkt.paritaLSB &= 0x01;
    } break;
  }
}

/*******************************************************************************
 *    Function: vypocitajParitu
 * Description: Spocitanie poctu bitov s hodnotou 1 v premennej n
 *   Parameter: [IN] uint32_t n - bity pre urcenie parity
 *      Return: uint8_t - pocet bitov s hodnotou 1
 ******************************************************************************/
/*static uint8_t vypocitajParitu( uint32_t n )
{
  n = (( n & 0xAAAAAAAA ) >> 1 ) + ( n & 0x55555555 );
  n = (( n & 0x30C30C30 ) >> 4 ) + (( n & 0x0C30C30C ) >> 2 ) + ( n & 0xC30C30C3 );
  return ( n % 63 );
}*/

/*******************************************************************************
 *    Function: vypocitajParitu
 * Description: Spocitanie poctu bitov s hodnotou 1 v premennej n
 *   Parameter: [IN] uint8_t n - bity pre urcenie parity
 *      Return: uint8_t - pocet bitov s hodnotou 1
 ******************************************************************************/
static uint8_t vypocitajParitu( uint8_t n )
{
  n = (( n & 0xAA ) >> 1 ) + ( n & 0x55 );
  n = (( n & 0xC0 ) >> 6 ) + (( n & 0x30 ) >> 4 ) + (( n & 0x0C ) >> 2 ) + ( n & 0x03 );
  //for ( uint8_t i = 0; i < 8; i++ )
  //  parita += ( n >> i ) & 0x01;
  return n;
}

/*******************************************************************************
 *    Function: posliBit
 * Description: Odoslanie bitu protokolom Wiegand
 *   Parameter: [IN] uint8_t vystup - vystup DATA0 alebo DATA1
 *      Return: Nič
 ******************************************************************************/
static void posliBit( uint8_t vystup)
{
  digitalWrite( vystup, LOW );
  delayMicroseconds( 34 );          // cas zotrvania v stave LOW - 50 us
  digitalWrite( vystup, HIGH );
  delayMicroseconds( 1000 );        // prestavka medzi jednotlivymi bitmi - 1000 us
}


FamilyCode.h
#ifndef FAMILY_CODE_H
#define FAMILY_CODE_H

#include <Arduino.h>

/*
Family Code   Part
() - iButton Package  Description
(Memory size in bits unless specified)
01 (hex)  (DS1990A), (DS1990R), DS2401, DS2411  1-Wire net address (registration number) only
02  (DS1991)¹   Multikey iButton, 1152-bit secure memory
04  (DS1994), DS2404  4Kb NV RAM memory and clock, timer, alarms
05  DS2405¹   Single addressable switch
06  (DS1993)  4Kb NV RAM memory
08  (DS1992)  1Kb NV RAM memory
09  (DS1982), DS2502  1Kb EPROM memory
0A  (DS1995)  16Kb NV RAM memory
0B  (DS1985), DS2505  16Kb EPROM memory
0C  (DS1996)  64Kb NV RAM memory
0F  (DS1986), DS2506  64Kb EPROM memory
10  (DS1920)  Temperature with alarm trips
12  DS2406, DS2407¹   1Kb EPROM memory, 2-channel addressable switch
14  (DS1971), DS2430A¹  256-bit EEPROM memory and 64-bit OTP register
1A  (DS1963L)¹  4Kb NV RAM memory with write cycle counters
1C  DS28E04-100   4096-bit EEPROM memory, 2-channel addressable switch
1D  DS2423¹   4Kb NV RAM memory with external counters
1F  DS2409¹   2-channel addressable coupler for sub-netting
20  DS2450  4-channel A/D converter (ADC)
21  (DS1921G), (DS1921H), (DS1921Z)   Thermochron® temperature logger
23  (DS1973), DS2433  4Kb EEPROM memory
24  (DS1904), DS2415  Real-time clock (RTC)
27  DS2417  RTC with interrupt
29  DS2408  8-channel addressable switch
2C  DS2890¹   1-channel digital potentiometer
2D  (DS1972), DS2431  1024-bit, 1-Wire EEPROM
37  (DS1977)  Password-protected 32KB (bytes) EEPROM
3A  (DS2413)  2-channel addressable switch
41  (DS1922L), (DS1922T), (DS1923), DS2422  High-capacity Thermochron (temperature) and Hygrochron™ (humidity) loggers
42  DS28EA00  Programmable resolution digital thermometer with sequenced detection and PIO
43  DS28EC20  20Kb 1-Wire EEPROM
*/

//            iBUTTON   FAMILY CODE   POPIS ( Pokial nie je uvedene inak, velkost pamate je v bitoch)
const uint8_t DS1990A = 0x01;         // DS1990R, DS2401, DS2411  1-Wire net address (registration number) only
const uint8_t DS1991  = 0x02;         // Multikey iButton, 1152-bit secure memory



#endif  // #ifndef FAMILY_CODE_H
